docs/slices.md

```diff
@@ -0,0 +1,1881 @@
+   in the direction being sliced. Slices like `a[:i]` or `a[i:]` should be
+   though of as the `start` or `stop` being omitted, not as a colon to the
+   left or right of an index. (See section {ref}`omitted`)
+7. Slicing something never raises an `IndexError`, even if the slice is empty.
+   For a NumPy array, a slice always keeps the axis being sliced, even if that
+   means the resulting dimension will be 0 or 1. (See section {ref}`subarray`)
+
+(integer-indices)=
+## Integer indices
+
+To understand slices, it is good to first review how integer indices work.
+Throughout this guide, we will use as an example this prototype list:
+
+$$
+a = [\mathtt{\textsf{'}a\textsf{'}},\ \mathtt{\textsf{'}b\textsf{'}},\ \mathtt{\textsf{'}c\textsf{'}},\
+\mathtt{\textsf{'}d\textsf{'}},\ \mathtt{\textsf{'}e\textsf{'}},\ \mathtt{\textsf{'}f\textsf{'}},\ \mathtt{\textsf{'}g\textsf{'}}]
+$$
+
+The list `a` has 7 elements.
+
+The elements of `a` are strings, but the indices and slices on `a` will always
+use integers. An index or slice is never based on the value of the elements,
+but rather the position of the elements in the list.[^dict-footnote]
+
+[^dict-footnote]: If you are looking for something that allows non-integer
+indices or that indexes by value, you may want a `dict`. Despite using similar
+syntax, `dict`s do not allow slicing.
+
+An integer index picks a single element from the list `a`. For NumPy arrays,
+integer indices pick a subarray corresponding to a particular element from a
+given axis (and as a result, an integer index always reduces the
+dimensionality of an array by one).
+
+(fourth-sentence)=
+The key thing to remember about indexing in Python, both for integer and
+slice indexing, is that it is 0-based. This means that the indices start
+at 0. This is the case for all **nonnegative** indices.
+For example, `a[3]` would pick the **fourth** element of `a`, in this case, `'d'`.
+
+<div style="text-align:center">
+<code style="font-size: 16pt;">a[3] == 'd'</code>
+$$
+\begin{aligned}
+\begin{array}{r c c c c c c c}
+a = & [\mathtt{\textsf{'}a\textsf{'}}, & \mathtt{\textsf{'}b\textsf{'}}, & \mathtt{\textsf{'}c\textsf{'}}, & \mathtt{\textsf{'}d\textsf{'}}, & \mathtt{\textsf{'}e\textsf{'}}, & \mathtt{\textsf{'}f\textsf{'}}, & \mathtt{\textsf{'}g\textsf{'}}]\\
+\color{red}{\text{index}}
+    & \color{red}{0\phantom{,}}
+    & \color{red}{1\phantom{,}}
+    & \color{red}{2\phantom{,}}
+    & \color{blue}3{\phantom{,}}
+    & \color{red}{4\phantom{,}}
+    & \color{red}{5\phantom{,}}
+    & \color{red}{6\phantom{,}}\\
+\end{array}
+\end{aligned}
+$$
+</div>
+
+```py
+>>> a = ['a', 'b', 'c', 'd', 'e', 'f', 'g']
+>>> a[3]
+'d'
+```
+
+0-based indexing is different from how people typically count things, which is
+1-based (1, 2, 3, ...). Thinking in terms of 0-based indexing requires some
+practice.
+
+For **negative** integers, indices index from the end of the list. These
+indices are necessarily 1-based (or rather, −1-based), since `0` already
```

Comment:
I always think of the indexes lying between the elements, rather than on them.  0 then refers to either end of the list

Reply:
And slicing is then an operation that sweeps through the specified indexes

Reply:
ie the indexes live with the commas and the brackets

Reply:
Oh I guess you address this below

Reply:
Yeah, I'm not a fan of this way of thinking of things.